br/pkg/restore/split/split.go

```diff
@@ -288,34 +288,88 @@ func (b *BackoffMayNotCountBackoffer) Attempt() int {
 }
 
 // GetSplitKeysOfRegions checks every input key is necessary to split region on
-// it. Returns a map from original region ID to split keys belongs to each
-// region.
+// it. Returns a map from region to split keys belongs to it.
+//
+// The key will be skipped if any of the following conditions is met:
+//   - it's the region boundary.
+//   - the key ranges consist of [prevKey, thisKey), [thisKey, nextKey) is already
+//     split.
 //
 // prerequisite:
 // - sortedKeys are sorted in ascending order.
 // - sortedRegions are continuous and sorted in ascending order by start key.
 // - sortedRegions can cover all keys in sortedKeys.
 // PaginateScanRegion should satisfy the above prerequisites.
-func GetSplitKeysOfRegions(sortedKeys [][]byte, sortedRegions []*RegionInfo, isRawKV bool) map[uint64][][]byte {
-	splitKeyMap := make(map[uint64][][]byte, len(sortedRegions))
+func GetSplitKeysOfRegions(
+	sortedKeys [][]byte,
+	sortedRegions []*RegionInfo,
+	isRawKV bool,
+) map[*RegionInfo][][]byte {
+	splitKeyMap := make(map[*RegionInfo][][]byte, len(sortedRegions))
+	pendingSplitKeyMap := make(map[*RegionInfo][][]byte, 1)
 	curKeyIndex := 0
+	splitKey := codec.EncodeBytesExt(nil, sortedKeys[curKeyIndex], isRawKV)
+	moveRegionCnt := 0
+
 	for _, region := range sortedRegions {
-		for ; curKeyIndex < len(sortedKeys); curKeyIndex += 1 {
+		for {
 			if len(sortedKeys[curKeyIndex]) == 0 {
-				continue
+				// should not happen?
+				goto nextKey
 			}
-			splitKey := codec.EncodeBytesExt(nil, sortedKeys[curKeyIndex], isRawKV)
 			// If splitKey is the boundary of the region, don't need to split on it.
 			if bytes.Equal(splitKey, region.Region.GetStartKey()) {
-				continue
+				goto nextKey
 			}
-			// If splitKey is not in a region, we should move to the next region.
+			// If splitKey is not in this region, we should move to the next region.
 			if !region.ContainsInterior(splitKey) {
+				moveRegionCnt++
 				break
 			}
-			regionID := region.Region.GetId()
-			splitKeyMap[regionID] = append(splitKeyMap[regionID], sortedKeys[curKeyIndex])
+
+			//   key1      key2    key3
+			//    |    a    |    b  |
+			// ---+----+----+----+--+-----------
+			// region1 | region2 | region3 | region4
+			//
+			// key2 don't need to be split on. Considering the original region, we can have
```

Comment:
if data cluster around key2, we would have a large region

Reply:
based on this, seems we don't need split on `15` for this:
regions: (, 10), [10, 100), [100, 200), [200,)
keys:  11, 15, 111, 121

Reply:
> if data cluster around key2, we would have a large region

yes, but I'm just porting the old logic. I'll check it and provide the reference to the old logic soon

> we don't need split on 15 for this

Current rule is we skip 15 only if [11, 15) and [15, 111) are both split. But [11, 15) is not. Even if data is clustered around 15 like they are mostly located in [14.9, 15.1), but because the upper layer choose 15 it means [11, 15) and [15, 111) and ~96MB data so we should also split 15

Reply:
yes, but the regions already splitted on `100`, it met the case described in the comment

Reply:
[11, 15) is not split, nearest region is 10 and 100 which are not inside [11, 15)

Reply:
>  I'll check it and provide the reference to the old logic soon

I found I understand the old logic wrong 😂 In old logic, we only skip the ranges which are the exactly same as the region.
https://github.com/pingcap/tidb/blob/8ba1fa452b1ccdbfb85879ea94b9254aabba2916/br/pkg/lightning/backend/local/localhelper.go#L150-L165

I'll update this PR